rfcs/xxxx-collage.md

```diff
@@ -0,0 +1,863 @@
+  sub-graph for further compilation) or iii) moved into the standard Relay optimization passes run before
+  `CollageFuseOps`.
+- **BYOC-specific patterns must match after standard Relay passes**: We currently run `CollageFuseOps` just before
+  `FuseOps`, which means it runs after quite a number of standard Relay passes which themselves may invoke some
+  target-specific rewrites (eg for quantization legalization). However, currently BYOC `partition_for_<toolchain>`
+  functions run before any other Relay passes, and thus the Relay model is seen in it's 'clean' state. We need to
+  figure out how to move `CollageFuseOps` earlier in the pipeline to be before any target-specific rewrites or any 
+  passes which may make BYOC patterns and lowering more difficult, but without interfering with TVM's fusion. This
+  needs more thought.
+- **Higher tuning cost**: Obviously Collage needs to estimate the latency of many more candidate kernels, and each
+  candidate may itself trigger tuning during lowering. For TVM this can require O(thousands) of trials and take O(hours)
+  , so we'll be very dependent on cached tuning logs to amortize this cost between models for the same target.
+  Currently Collage will measure more candidates even if TVM native is the only available target.
+- **Task extraction vs Tuning**: Traditionally TVM has had three phases: i) Task extraction (find the fused sub-graphs
+  to tune), ii) Tuning (find a good schedule for those sub-graphs), and iii) Compilation (re-compile the model, now
+  retrieving schedules for all the anticipated sub-graphs from the cache.) However the Collage 'v2' prototype collapses
+  all these phases. This lets us lazily explore the implied search graph (nodes = partially rewritten models, edges =
+  selected of sub-graph and toolchain as a candidate kernel, cost = estimated sum of kernel costs plus launch penalties)
+  , and thus only pay the cost of tuning candidate kernels which could possibly influence the final partitioning.
+- **No non-local optimization**: Though Collage can explore the choice of sub-graph and toolchain, it cannot explore any
+  choices which require the arguments and/or result of the sub-graph to be rewritten, or the overall `IRModule` to be
+  changed. Thus Collage **cannot** be used to search over:
+    - choice of layout for arguments/results (may require insertion of layout transforms),
+    - choice of memory scope for arguments/results (may require insertion of device copies),
+    - choice of device on which to host the kernel (ditto),
+    - choice of quantization scheme,
+  
+  since all those choices can require changes beyond the candidate's sub-graph. To support this efficiently we'd need
+  to abandon the simple-minded but fast `SubGraph` representation we describe here in favor of something like an EGraph
+  representation, which seems like a very large change for TVM.
```

Comment:
I think your 'Plan' matches my 'CandidatePartition', which has a 'SubGraph'. I think it would be pretty straightforward for SubGraphs to be able to represent things like 'oh, by the way, this partition is on device X so needs some device copies on the way in and out'. But I remain skeptical that will do what we want unless we somehow account for post-partitioning optimizations to remove redundant copies etc.